### PR TITLE
feat: reconcile private hosts for AKS BYO DNS

### PR DESCRIPTION
**Pull request by levimm**
**Reason for Change**:

This pr is required by AKS private cluster BYO DNS scenario. Since this pr also changes the api model, the pr need to be merged into master.
For AKS BYO DNS, we'll config the /etc/hosts file on all agent nodes so we can resolve the private dns without azure private dns zone.

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: parts/k8s/cloud-init/nodecustomdata.yml [hunk 1]
```diff
@@ -328,6 +328,71 @@ write_files:
...
+{{if and IsHostedMaster EnableHostsConfigAgent}}
+- path: /opt/azure/containers/reconcilePrivateHosts.sh
+  permissions: "0744"
+  owner: root
+  content: |
+    #!/usr/bin/env bash
+    set -o nounset
+    set -o pipefail
+
+    clusterFQDN="test.privatelink.eastus.azmk8s.io"
```
Comment (feiskyer) [inline parts/k8s/cloud-init/nodecustomdata.yml:340]: remove the test FQDN here?

Comment (levimm) [reply parts/k8s/cloud-init/nodecustomdata.yml]: removed

**Comment by levimm** [comment]
Thanks @devigned for the comment. Rebased the commit and fixed API doc. Please help do a re-review and trigger the test.

**Pull request merged**